Assets/Mirror/Runtime/SyncDictionary.cs

```diff
@@ -0,0 +1,331 @@
+        }
+
+        public void OnDeserializeDelta(NetworkReader reader)
+        {
+            // This list can now only be modified by synchronization
+            IsReadOnly = true;
+
+            int changesCount = reader.ReadInt32();
+
+            for (int i = 0; i < changesCount; i++)
+            {
+                Operation operation = (Operation)reader.ReadByte();
+
+                // apply the operation only if it is a new change
+                // that we have not applied yet
+                bool apply = changesAhead == 0;
+                B key = default;
+                T item = default;
+
+                switch (operation)
+                {
+                    case Operation.OP_ADD:
+                        key = DeserializeKey(reader);
+                        item = DeserializeItem(reader);
+                        if (apply)
+                        {
+                            m_Objects.Add(key, item);
+                        }
+                        break;
+
+                    case Operation.OP_CLEAR:
+                        if (apply)
+                        {
+                            m_Objects.Clear();
+                        }
+                        break;
+
+                    case Operation.OP_REMOVE:
+                        key = DeserializeKey(reader);
+                        if (apply)
+                        {
+                            m_Objects.Remove(key);
+                        }
+                        break;
+
+                    case Operation.OP_SET:
+                    case Operation.OP_DIRTY:
+                        key = DeserializeKey(reader);
+                        item = DeserializeItem(reader);
+                        if (apply)
+                        {
+                            m_Objects[key] = item;
+                        }
+                        break;
+                }
+
+                if (apply)
+                {
+                    Callback?.Invoke(operation, key, item);
+                }
+                // we just skipped this change
+                else
+                {
+                    changesAhead--;
+                }
+            }
+        }
+
+
+        public void Clear()
+        {
+            m_Objects.Clear();
+            AddOperation(Operation.OP_CLEAR, default);
+        }
+
+        public bool ContainsKey(B key) => m_Objects.ContainsKey(key);
+
+        public bool Remove(B key)
+        {
+            bool result = m_Objects.Remove(key);
+            if (result)
+            {
+                AddOperation(Operation.OP_REMOVE, key);
+            }
+            return result;
+        }
+
+        public void Dirty(B index)
+        {
+            AddOperation(Operation.OP_DIRTY, index, m_Objects[index]);
+        }
+
+        public T this[B i]
+        {
+            get { return m_Objects[i]; }
+            set
+            {
+                bool changed = false;
+                if (m_Objects[i] == null)
```

Comment:
This will throw KeyNotFound exception  if you are trying to set a key that is already in the dictionary